File: make_inputs.py
import datetime
import logging
import os.path
import sys
from math import exp, log, pow, sqrt

import pandas as pd

import util
from util import save_csv

logger = logging.getLogger("cffdrs")
logger.setLevel(logging.WARNING)

# HOUR_TO_START_FROM = 12

# Fuel Load (kg/m^2)
DEFAULT_GRASS_FUEL_LOAD = 0.35

# FIX: figure out what this should be
DEFAULT_LATITUDE = 55.0
DEFAULT_LONGITUDE = -120.0


if "__main__" == __name__:
    # CSV headers
    header = "lat,long,yr,mon,day,hr,temp,rh,ws,prec"
    header_out = (
        "lat,long,yr,mon,day,hr,temp,rh,ws,prec,solrad,percent_cured,grass_fuel_load"
    )
    if 4 != len(sys.argv):
        logger.fatal(
            "\n".join(
                [
                    f"Command line:   {sys.argv[0]}  <local GMToffset> <input file> <output file>\n",
                    "<local GMToffset> is the off of Greenich mean time (for Eastern = -5  Central=-6   MT=-7  PT=-8 )",
                    "All times should be local standard time",
                    "INPUT FILE format must be HOURLY weather data, comma separated and take the form",
                    f"{header}\n",
                ]
            )
        )
        sys.exit(1)
    outfile = sys.argv[3]
    infile = sys.argv[2]
    if not os.path.exists(infile):
        logger.fatal(f"/n/n ***** FILE  {infile}  does not exist\n")
        sys.exit(1)
    timezone = int(sys.argv[1])
    if timezone < -9 or timezone > -2:
        logger.fatal(
            "/n *****   Local time zone adjustment must be vaguely in CAnada so between -9 and -2"
        )
        sys.exit(1)
    # colnames_in = ["lat", "long", "year", "mon", "day", "hour", "temp", "rh", "wind", "rain"]
    # df = pd.read_csv(infile, header=None, names=colnames_in)
    wx = pd.read_csv(infile)
    logger.debug(wx)
    # solar radiation function relies on all hours of the day, so need to pass
    # whole frame
    # print(solrad, sunrise, sunset)
    if any(wx.columns != header.split(",")):
        raise RuntimeError(f"Expected columns {header}")
    wx.columns = [x.upper() for x in wx.columns]
    # require exact columns for now
    if "MINUTE" not in wx.columns:
        wx.loc[:, "MINUTE"] = 0
    # print(wx.dtypes)
    if "TIMESTAMP" not in wx.columns:
        wx.loc[:, "TIMESTAMP"] = wx.apply(
            lambda row: datetime.datetime(
                int(row["YR"]),
                int(row["MON"]),
                int(row["DAY"]),
                int(row["HR"]),
                int(row["MINUTE"]),
            ),
            axis=1,
        )
    wx["TIMEZONE"] = timezone
    wx = util.get_sunlight(wx, with_solrad=True)
    if "PERCENT_CURED" not in wx.columns:
        wx["JULIAN"] = wx.apply(lambda row: util.julian(row["MON"], row["DAY"]), axis=1)
        wx["PERCENT_CURED"] = wx["JULIAN"].apply(util.seasonal_curing)
    if "GRASS_FUEL_LOAD" not in wx.columns:
        wx["GRASS_FUEL_LOAD"] = DEFAULT_GRASS_FUEL_LOAD
    wx.columns = [x.lower() for x in wx.columns]
    colnames_out = header_out.split(",")
    wx = wx.loc[:, colnames_out]
    save_csv(wx, outfile)

File: make_inputs.r
#' Computes hourly FWI indices for an input hourly weather stream
library(lubridate)
library(data.table)
source("util.r")

# Fuel Load (kg/m^2)
DEFAULT_GRASS_FUEL_LOAD <- 0.35

# FIX: figure out what this should be
DEFAULT_LATITUDE <- 55.0
DEFAULT_LONGITUDE <- -120.0


#' Calculate hourly FWI indices from hourly weather stream.
#'
#' @param     df_wx           hourly values weather stream
#' @param     timezone        integer offset from GMT to use for sun calculations
#' @return                    hourly inputs for FWI calculation
#' @export make_inputs
make_inputs <- function(df_wx, timezone) {
    wx <- as.data.table(copy(df_wx))
    # only return specific columns - don't worry about other columns right now
    header <- "lat,long,yr,mon,day,hr,temp,rh,ws,prec"
    header_out <- "lat,long,yr,mon,day,hr,temp,rh,ws,prec,solrad,percent_cured,grass_fuel_load"
    cols_req <- unlist(strsplit(header, ","))
    stopifnot(colnames(wx) == cols_req)
    colnames(wx) <- toupper(colnames(wx))
    wx[, MINUTE := 0]
    wx[, TIMESTAMP := as_datetime(sprintf("%04d-%02d-%02d %02d:%02d:00", YR, MON, DAY, HR, MINUTE))]
    wx[, DATE := as.Date(TIMESTAMP)]
    wx$JULIAN <- julian(wx$MON, wx$DAY)
    wx$PERCENT_CURED <- seasonal_curing(wx$JULIAN)
    wx$GRASS_FUEL_LOAD <- DEFAULT_GRASS_FUEL_LOAD
    wx$TIMEZONE <- timezone
    wx <- get_sunlight(wx, with_solrad = TRUE)
    cols_out <- unlist(strsplit(header_out, ","))
    colnames(wx) <- tolower(colnames(wx))
    wx <- wx[, ..cols_out]
    return(wx)
}

# so this can be run via Rscript
if ("--args" %in% commandArgs()) {
    args <- commandArgs(trailingOnly = TRUE)
    if (3 == length(args)) {
        # args <- c("-6", "./out/wx_diurnal_r.csv", "./out/inputs_diurnal_r.csv")
        timezone <- as.double(args[1])
        file_in <- args[2]
        file_out <- args[3]
        df_wx <- as.data.table(read.csv(file_in))
        df_inputs <- make_inputs(df_wx, timezone = timezone)
        save_csv(df_inputs, file_out)
    } else {
        message("Wrong number of arguments")
    }
}
